src/algebra/monoid_algebra/degree.lean

```diff
@@ -0,0 +1,161 @@
+end semiring
+
+lemma sup_support_pow_le [comm_semiring R] [add_monoid A] [add_monoid B]
+  [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
+  {D : A → B} (D0 : D 0 ≤ 0) (Dm : ∀ a b, D (a + b) ≤ D a + D b) (n : ℕ)
+  (f : add_monoid_algebra R A) :
+  (f ^ n).support.sup D ≤ n • (f.support.sup D) :=
+begin
+  induction n with n hn,
+  { simp only [pow_zero, zero_smul, finset.sup_le_iff],
+    exact λ a ha, by rwa finset.mem_singleton.mp (finsupp.support_single_subset ha) },
+  { rw [pow_succ, succ_nsmul],
+    exact (sup_support_mul_le Dm _ _).trans (add_le_add rfl.le hn) }
+end
+
+variables [comm_semiring R] [add_comm_monoid A] [add_comm_monoid B]
+  [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
+
+lemma sup_support_multiset_prod_le
+  {D : A → B} (D0 : D 0 ≤ 0) (Dm : ∀ a b, D (a + b) ≤ D a + D b)
+  (F : multiset (add_monoid_algebra R A)) :
+  F.prod.support.sup D ≤ (F.map (λ f : add_monoid_algebra R A, f.support.sup D)).sum :=
+begin
+  induction F using quot.induction_on,
+  rw [multiset.quot_mk_to_coe'', multiset.coe_map, multiset.coe_sum, multiset.coe_prod],
+  exact sup_support_list_prod_le D0 Dm F,
+end
+
+lemma sup_support_finset_prod_le {ι : Type*}
+  {D : A → B} (D0 : D 0 ≤ 0) (Dm : ∀ a b, D (a + b) ≤ D a + D b)
+  (s : finset ι) (f : ι → add_monoid_algebra R A):
+  (∏ i in s, f i).support.sup D ≤ ∑ i in s, (f i).support.sup D :=
+begin
+  refine (sup_support_multiset_prod_le D0 Dm _).trans (le_of_eq _),
+  simp only [multiset.map_map, function.comp_app],
+  refl,
+end
+
+end general_results_assuming_semilattice_sup
+
+section degrees
+
+variables [semiring R]
+
+section degree
+
+variables [semilattice_sup A]
+
+/--  Let `R` be a semiring, let `A` be a `semilattice_sup`, and let `f : R[A]` be an
+an element of `add_monoid_algebra R A`.  The degree of `f` takes values in `with_bot A` and
+it is the supremum of the support of `f` or `⊥`, depending on whether `f` is non-zero or not.
+
+If `A` has a linear order, then this notion coincides with the usual one, using the maximum of
+the exponents. -/
+@[reducible]
+def degree (f : add_monoid_algebra R A) : with_bot A :=
+f.support.sup coe
+
+variables [add_monoid A] [covariant_class A A (+) (≤)] [covariant_class A A (function.swap (+)) (≤)]
+variables (f g : add_monoid_algebra R A)
+
+lemma degree_mul_le : (f * g).degree ≤ f.degree + g.degree :=
+sup_support_mul_le (λ a b, (with_bot.coe_add _ _).le) f g
+
+end degree
+
+section trailing_degree
+
+variables [semilattice_inf A]
+
+/--  Let `R` be a semiring, let `A` be a `semilattice_inf`, and let `f : R[A]` be an
+an element of `add_monoid_algebra R A`.  The trailing degree of `f` takes values in `with_top A`
+and it is the infimum of the support of `f` or `⊤`, depending on whether `f` is non-zero or not.
+
+If `A` has a linear order, then this notion coincides with the usual one, using the minimum of
+the exponents. -/
+def trailing_degree (f : add_monoid_algebra R A) : with_top A :=
```

Comment:
I do not know if there is a standard terminology.  I like `max_degree` and `min_degree`.  Techincally, `sup_degree` and `inf_degree` would be more correct, but I think that they convey the wrong idea that there is actually a sup/inf-that-is-not-a-max, rather than explaining that there is an ambiguity with top/bot.

I am in favour of renaming them `max/min_degree`.  This way, when we relate them to the usual `polynomial.degree` we can name the lemma `degree_eq_max_degree`, instead of the awkward `degree_eq_degree`. :upside_down_face: